### PR TITLE
Management API: Generate valid operation IDs for minor API versions, and upgrade @hey-api/openapi-ts

### LINKED ISSUE #23429: Vunlerable dependencies in Umbraco.Web.Ui.Client\package.json
### Which Umbraco version are you using?

release/17.6.0-rc

### Bug summary

`npm install` shows following result:
```

up to date, audited 1246 packages in 6s

329 packages are looking for funding
  run `npm fund` for details

16 vulnerabilities (3 low, 6 moderate, 6 high, 1 critical)

To address issues that do not require attention, run:
  npm audit fix

To address all issues (including breaking changes), run:
  npm audit fix --force

Run `npm audit` for details.
npm warn install-scripts 2 packages had install scripts blocked because they are not covered by allowScripts:
npm warn install-scripts   esbuild@0.27.7 (postinstall: node install.js)
npm warn install-scripts   msw@2.14.6 (postinstall: node -e "import('./config/scripts/postinstall.js').catch(() => void 0)")
npm warn install-scripts
npm warn install-scripts Run `npm install-scripts ls` to review, or `npm install-scripts approve <pkg>` to allow.
```

### Specifics

_No response_

### Steps to reproduce

E.g. checkout branch `release/17.6` open folder `src\Umbraco.Web.Ui.Client` in VS Code and execute `npm i` in terminal.

### Expected result / actual result

No severities with high or critical should be listed - I guess😉

### PR DESCRIPTION
**Pull request by iOvergaard**
Fixes the Management API's operation ID generation for minor API versions, and upgrades `@hey-api/openapi-ts` to 0.99.0 now that the fix unblocks it.

Fixes #23429 in part — see the note at the bottom.

## The bug

The `v1.1` document validate endpoint generated this operation ID:

```
PutUmbracoManagementApiV1.1DocumentByIdValidate1.1
```

Dots in it, and the whole route prefix leaked in — compare its siblings, `PutMediaByIdValidate` and `PutMemberByIdValidate`. Two defects in `OperationIdHandler` combine:

1. `VersionPrefixRegex` was `.*?\/v[1-9]+/`, which doesn't match `/v1.1/`, so the prefix was never stripped. It didn't match `/v10/` either, since `0` isn't in `[1-9]`.
2. The `MapToApiVersion` value was appended verbatim, putting a second dot on the end.

`v1.1` is currently the only minor-versioned Management endpoint, which is why this went unnoticed. Dots aren't valid in generated client identifiers, and `@hey-api/openapi-ts` splits operation IDs on `/[./]/` to derive nesting — so 0.85 mangled them into `…Validate11`, while newer versions split on them and emitted a container holding a method literally named `1`.

## The fix

- **Widened `VersionPrefixRegex`** to match minor and multi-digit versions.
- **Substituted rather than removed** the invalid characters in the version suffix, so `1.1` becomes `1_1` and cannot collide with a genuine `11` version (or `1.11` with `11.1`).
- **Scoped that substitution to the Management API.** This handler also serves the Delivery API, whose `2.0` endpoints have published IDs like `GetContent2.0`; all 8 of them were being renamed, which is a breaking change for headless consumers generating clients from the Delivery spec. There's a `TODO (V19)` to apply it there as a documented breaking change.
- **Pinned the v1.1 validate endpoint's ID** with `Name = "PutDocumentByIdValidate"`, so the generated client method stays version-free.

The version *prefix* change needed no scoping: Delivery routes are `/v2/`, which matched the old pattern as well as the new one, so only the suffix ever reached Delivery.

### On the pinned route name

The backoffice always calls the latest version of an endpoint, so a version embedded in the client method name carries no information — and without a pinned ID, every version bump renames the method and churns every call site. Pinning decouples the client from the version.

The convention is now documented in the Management API's `CLAUDE.md`, including two rules worth stating explicitly: only the **latest** version takes the unversioned name, and two versions must never share an ID, since duplicates break OpenAPI generation.

### A bug found while widening the tests

The missing-relative-path branch built its message from `RouteValues["controller"]` using the indexer, so anyone reaching it got a `KeyNotFoundException` instead of the intended `InvalidOperationException`. Now read defensively.

## @hey-api/openapi-ts 0.85.2 → 0.99.0

Unblocked by the above. `asClass`/`classNameBuilder` are deprecated, so they're migrated to the new `operations` config.

**The upgrade itself changes no symbols** — regenerating with 0.99 instead of 0.85 produces an identical list of 516 exported classes and methods, verified by diffing them.

The *operation ID fix* does change the surface, deliberately. Against `v17/dev` the net effect is three junk symbols removed and one clean one added:

```diff
+ DocumentService.putDocumentByIdValidate
- DocumentService...putUmbracoManagementApiV11DocumentByIdValidate11
- class DocumentByIdValidate1Service
- class PutUmbracoManagementApiV1Service   // was empty — a phantom container from the malformed ID
```

One code change was needed for the upgrade: the fetch client now types request results as `Response | undefined`, so the 401 retry path throws when no response came back, which the existing `.catch(req.reject)` already routes to the caller.

This raises the published peer floor from `>=0.85.0 <1.0.0` to `>=0.99.0 <1.0.0` for plugin developers.

### Be aware: this is a security trade, not a win

| | before | after |
|---|---|---|
| backoffice client | 1 critical, 4 high | **0 critical**, 7 high |

It drops the `handlebars` critical, but `@hey-api/json-schema-ref-parser` pins `js-yaml` at exactly `4.2.0`, which carries a high-severity DoS advisory that cascades onto `@hey-api/shared` and `openapi-ts` itself. **No release of openapi-ts is currently free of both** — everything `<=0.97.2` has handlebars, everything `>=0.97.1` has the js-yaml chain. `js-yaml@4.3.0` is clean but the pin is exact, and an npm `overrides` entry doesn't apply through the workspace nesting.

> **Post-merge update.** The client is now at **1 low, 2 moderate, 4 high, 0 critical** (7 total), down from the 7 high quoted above. The `brace-expansion` → `minimatch` → `eslint-plugin-import` chain (3 highs) cleared on its own once upstream published a patched `brace-expansion@1.1.18`, which the existing `^1.1.7` range picked up — `eslint-plugin-import` itself is unchanged. All 4 remaining highs are the single `js-yaml` pin described above. Running totals and the full breakdown are on [#23429](https://github.com/umbraco/Umbraco-CMS/issues/23429#issuecomment-5179890530).

All of it is build-time codegen that never reaches the browser. This should resolve upstream when json-schema-ref-parser bumps its pin.

For the **login app** this bump is only a stopgap: #23501 removes `@hey-api/openapi-ts` from it entirely, taking it to zero advisories rather than trading one for another.

## Testing

- `OperationIdHandlerTests` — **12 tests** (this class had no coverage before). A `TestCaseSource` covers the minor-version substitution, multi-digit versions, the default-version case, both branches of the explicit route name, and the Delivery scoping (pinning `GetContent2.0` so it can't silently regress). @AndyButland added `CanHandle` coverage across the Management, Delivery and out-of-namespace cases, and split the non-controller case so `CanHandle` returning `false` is asserted separately from `Handle` throwing `ArgumentException`. A further test covers the `InvalidOperationException` path
- **Both** OpenAPI contract tests pass — `NewBackoffice.OpenAPIContractTest` and `DeliveryApi.OpenApiContractTest`
- Frontend `build`, `lint:errors`, `check:circular` clean; `npm test` — **2587 passed, 0 failed**

## Notes

- A TipTap branch was accidentally merged into this one and has been reverted. The branch's net change to those files against the merge base is zero, and #23497 has since merged, so `v17/dev`'s version is what applies.
- This does **not** close #23429 on its own. The remaining advisories there are two upstream-gated chains — see [my comment on the issue](https://github.com/umbraco/Umbraco-CMS/issues/23429#issuecomment-5116590498).

---
🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by iOvergaard** [comment]
Thanks — both review points were valid and are fixed in dba315a:

- **Dead null check removed.** Confirmed: `GetMapToApiVersionAttributeValue()` returns a non-nullable `string` via `string.Join`, so the `is null` branch was unreachable. Simplified to the single `Replace` call.
- **Tests renamed** to `Can_Strip_Route_Prefix_Of_Minor_Api_Version`, `Can_Strip_Route_Prefix_Of_Multi_Digit_Api_Version` and `Can_Omit_Version_Suffix_For_Default_Api_Version`. The convention is real, though it lives in `.claude/skills/umb-review/references/coding-preferences.md` rather than a `CLAUDE.md`.

Tests still pass, and the two bug-specific ones still fail without the handler fix.

One heads-up on scope: the login half of this PR is superseded by #23501, which removes `@hey-api/openapi-ts` from the login project altogether (backporting #22757) rather than bumping it. That takes the login app to **0 advisories** instead of trading the `handlebars` critical for a `js-yaml` high. If #23501 lands, the login bump here becomes redundant.

File: src/Umbraco.Cms.Api.Common/OpenApi/OperationIdHandler.cs [hunk 1]
```diff
@@ -102,7 +102,9 @@ protected string UmbracoOperationId(ApiDescription apiDescription)
         // We only want to add a version, if it is not the default one.
         if (string.Equals(versionAttributeValue, defaultVersion.ToString()) == false)
         {
-            version = versionAttributeValue;
+            // Strip any characters that are not valid in an identifier, so a minor version like "1.1"
```
Comment (AndyButland) [inline src/Umbraco.Cms.Api.Common/OpenApi/OperationIdHandler.cs]: As this is in the "Common" API project, it'll apply also to the Delivery API, and so also rewrite the Delivery API's 2.0 suffix.  So e.g. `GetContent2.0` becomes `GetContent20`, which you can see if you run the, now failing, `OpenApiContractTest` in `Umbraco.Cms.Tests.Integration.Umbraco.Core.DeliveryApi`.

Arguably it's correct to fix this for the Delivery API too, but seems to me we'd need to wait for a major to do this, otherwise headless users generating clients from the Delivery spec would get renamed operations.

It may be best to target this for the management API - which I think you can do with a check for `controllerActionDescriptor.ControllerTypeInfo.Namespace?.StartsWith("Umbraco.Cms.Api.Management")` - as that only affects the existing 1.1. endpoints that, as you note, are generated invalid IDs anyway.

And add a `TODO` to remove the check for a documented breaking Delivery API change in 19.

Comment (AndyButland) [inline src/Umbraco.Cms.Api.Common/OpenApi/OperationIdHandler.cs]: Maybe replace rather than delete is better, so a `.` becomes an `_`.  Otherwise, theoretically you could get clashes - 1.1 and 11, 1.11 and 11.1, etc.

File: src/Umbraco.Cms.Api.Management/Controllers/Document/ValidateUpdateDocumentController.cs [hunk 2]
```diff
@@ -43,7 +43,7 @@ public ValidateUpdateDocumentController(
         _backOfficeSecurityAccessor = backOfficeSecurityAccessor;
     }
 
-    [HttpPut("{id:guid}/validate")]
+    [HttpPut("{id:guid}/validate", Name = "PutDocumentByIdValidate")]
```
Comment (AndyButland) [inline src/Umbraco.Cms.Api.Management/Controllers/Document/ValidateUpdateDocumentController.cs:48]: Do we need an explicit name here if we have fixed the operation IDs to generate a valid one?  Concern is that if we start doing this we run the risk of adding operation names that don't match the convention and end up with a bit of a mix.

File: tests/Umbraco.Tests.UnitTests/Umbraco.Cms.Api.Common/OpenApi/OperationIdHandlerTests.cs
Comment (AndyButland) [inline tests/Umbraco.Tests.UnitTests/Umbraco.Cms.Api.Common/OpenApi/OperationIdHandlerTests.cs:1]: Great that this is added, as this class wasn't under test before. But I can see it's only testing a subset of the functionality.  It's probably a good idea whilst introducing this test class to ensure the other branches of code in the class are tested too.

- The explicit route Name short-circuit (OperationIdHandler.cs:65-71) is untested (and currently used in this PR).
- The remaining surface looks like it should be straightforward to add tests for: `ArgumentException` on a non-`ControllerActionDescriptor`, `InvalidOperationException` on empty `RelativePath`, and the manipulation that follows.

**Review by AndyButland — CHANGES_REQUESTED** [review]
Looks good @iOvergaard and resolves the issue, but I have a concern about this leaking into the Delivery API and whether this could cause breaking changes for headless consumers.  Please see my comments inline.

**Comment by iOvergaard** [comment]
Thanks @AndyButland — all four addressed in 38e24d7. Taking them in turn:

**Delivery API leak.** You were right, and I'd missed it because I ran `NewBackoffice.OpenAPIContractTest` but not `DeliveryApi.OpenApiContractTest` — two similarly named contract tests, and I verified the half I happened to be thinking about. All **8** Delivery actions carry `MapToApiVersion("2.0")`, so every one was getting renamed.

Scoped to the Management API as you suggested, with the `TODO (V19)`. Both contract tests now pass.

Worth noting the version *prefix* change needed no scoping — Delivery routes are `/v2/`, which matched the old `[1-9]+` pattern as well as the new one, so only the suffix ever reached Delivery.

**Replace rather than delete.** Agreed, and the collision examples are fair — the spec now carries `PutDocumentByIdValidate1_1`.

One caveat you'll want to know: **openapi-ts normalises the underscore away**, generating `putDocumentByIdValidate11`. So the collision-safety holds for the spec and for generators that preserve it, but the hey-api client would still collide on a hypothetical `1.11` vs `11.1`. Still the right call — the spec is the published contract — but the protection doesn't reach all the way down.

**Explicit route name.** Dropped. Your convention-drift concern is the stronger argument, and with the generator fixed it's redundant. The ID is now `PutDocumentByIdValidate1_1` by convention.

**Test coverage.** Expanded from 3 to 8, covering everything you listed — both branches of the explicit-route-name short-circuit, the `ArgumentException`, the `InvalidOperationException` — plus a Delivery case that pins `GetContent2.0` so this scoping can't silently regress again. The namespace checks key off the real `ValidateUpdateDocumentController` and `QueryContentApiController` types rather than fakes, so the tests track the namespaces that actually ship.

And writing that coverage earned its keep immediately: the missing-relative-path branch built its message from `RouteValues["controller"]` using the indexer, so anyone hitting it got a `KeyNotFoundException` instead of the intended `InvalidOperationException`. Fixed to read defensively.

---

Verified: 8/8 unit tests, both OpenAPI contract tests, and the frontend suite at 2562 passed. The tiptap commits from the accidental merge are being left for #23497 to land, which will drop them from this diff on its own — reverting the merge here risks carrying that revert into `v17/dev` and undoing Lee's work.

**Comment by AndyButland** [comment]
Fixes all look good @iOvergaard.

---

I think you can revert the TipTap work I accidentally merged in here actually, rather than waiting for it to be cleared.  We'll squash this PR when it comes to merging, so neither the merge nor the revert would survives into `v17/dev`.

```
git revert -m 1 96140a040f6f12605a4955a3a777b435ec69e307
```

---

I'm questioning my comment from earlier now seeing this, around applying a custom `Name` in the API endpoint attribute.  It makes the client-side code a bit uglier now you have to deal with `DocumentService.putDocumentByIdValidate1_1`.  It seems unlikely that the backoffice client is ever going to have to be concerned about versions - it'll call whatever the latest endpoint is, so having the version embedded in the method name feels a bit superfluous.

What do you think? Maybe it's actually better to add a name on the 1.1 version so you remain calling `DocumentService.putDocumentByIdValidate(...`. And so reinstate `Name = "PutDocumentByIdValidate"` on the 1.1 endpoint, and document this in `CLAUDE.md` so it's clear what should be done when versioning a management API endpoint for use in the backoffice.

---

There are some merge conflicts that will need resolution before this can be merged.

**Review by AndyButland — APPROVED** [review]
Looks good to me now @iOvergaard.  I pushed a few more small commits, addressing test coverage and documentation, but if that all seems fine to you and the checks pass, please go ahead and merge.
